internal/localtypes/file_permission_type.go

```diff
@@ -0,0 +1,76 @@
+package localtypes
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr/xattr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+var (
+	_ types.StringTypable    = FilePermissionType{}
+	_ xattr.TypeWithValidate = FilePermissionType{}
+)
+
+type FilePermissionType struct {
+	types.StringTypable
+}
+
+func NewFilePermissionType() FilePermissionType {
+	return FilePermissionType{StringTypable: types.StringType}
+}
+
+// Validate checks that the given input string is a valid file permission string,
+// expressed in numeric notation.
+// See: https://en.wikipedia.org/wiki/File-system_permissions#Numeric_notation
+func (f FilePermissionType) Validate(ctx context.Context, value tftypes.Value, path path.Path) diag.Diagnostics {
```

Comment:
You could add `var diags diag.Diagnostics`  as the first line of this func then call `diags.Append(...)` below so that you don't have to instantiate diag.Diagnostics more than once.